aead/src/lib.rs

```diff
@@ -68,96 +75,269 @@ impl fmt::Display for Error {
-    #[cfg(feature = "getrandom")]
-    fn generate_nonce() -> core::result::Result<Nonce<Self>, getrandom::Error>
-    where
-        Nonce<Self>: Default,
-    {
-        let mut nonce = Nonce::<Self>::default();
-        getrandom::getrandom(&mut nonce)?;
-        Ok(nonce)
+    /// The returned byte slice is guaranteed to point to the output of `buffer`.
+    #[inline]
+    fn postfix_encrypt_to_buf<'out>(
+        &self,
+        nonce: &Nonce<Self>,
+        associated_data: &[u8],
+        plaintext: &[u8],
+        buffer: &'out mut [u8],
+    ) -> Result<&'out mut [u8]> {
+        let tag_len = Self::TagSize::USIZE;
+        let res_len = plaintext.len() + tag_len;
+        let buf = buffer.get_mut(..res_len).ok_or(Error)?;
+        let (msg_dst, tag_dst) = buf.split_at_mut(plaintext.len());
+        let inout_buf = InOutBuf::new(plaintext, msg_dst).expect("ct_dst has correct length");
+        let tag = self.detached_encrypt_inout(nonce, associated_data, inout_buf)?;
+        tag_dst.copy_from_slice(&tag);
+        Ok(buf)
     }
 
-    /// Generate a random nonce for this AEAD algorithm using the specified
-    /// [`CryptoRngCore`].
+    /// Decrypt the data in `ciphertext`, write resulting ciphertext to `buffer`, verify
+    /// the appended authentication tag, and return the resulting byte slice in case of success.
     ///
-    /// See [`AeadCore::generate_nonce`] documentation for requirements for
-    /// random nonces.
-    #[cfg(feature = "rand_core")]
-    fn generate_nonce_with_rng(
-        rng: &mut impl CryptoRngCore,
-    ) -> core::result::Result<Nonce<Self>, rand_core::Error>
-    where
-        Nonce<Self>: Default,
-    {
-        let mut nonce = Nonce::<Self>::default();
-        rng.try_fill_bytes(&mut nonce)?;
-        Ok(nonce)
+    /// Returns an error if the provided authentication tag does not match the given ciphertext,
+    /// if the size of `ciphertext` is smaller than the tag size, or if the size of `buffer` is
+    /// too small for resulting plaintext (i.e. it should have capacity of at least
+    /// `ciphertext.len() - tag_size`).
+    ///
+    /// The returned byte slice is guaranteed to point to the output of `buffer`.
+    #[inline]
+    fn postfix_decrypt_to_buf<'out>(
+        &self,
+        nonce: &Nonce<Self>,
+        associated_data: &[u8],
+        ciphertext: &[u8],
+        buffer: &'out mut [u8],
+    ) -> Result<&'out mut [u8]> {
+        let tag_len = Self::TagSize::USIZE;
+        let pt_len = ciphertext.len().checked_sub(tag_len).ok_or(Error)?;
+        let pt_dst = buffer.get_mut(..pt_len).ok_or(Error)?;
+        let (ct, tag) = ciphertext.split_at(pt_len);
+        let tag = tag.try_into().expect("tag has correct length");
+        let buf = InOutBuf::new(ct, pt_dst).expect("buffers have the same length");
+        self.detached_decrypt_inout(nonce, associated_data, buf, tag)?;
+        Ok(pt_dst)
+    }
+
```

Comment:
These impls are all incredibly redundant and suggest to me that this PR isn't working at a good abstraction level.

Reply:
The current implementation is based on abstractions which eliminate this boilerplate. Your version is very much not DRY.

Reply:
The current implementation does not deal with `inout`. I could've just removed the `inplace` and `to_buf` methods to comply with DRY, but I doubt users would like such API.

Reply:
The changes needed to support `inout` are minimal. I can continue them in #1714 (or a separate PR) for comparison.

Reply:
I have reduced the amount of code duplication in this [commit](https://github.com/RustCrypto/traits/pull/1713/commits/1e55bec6044b5c32a45143024a25b24f2064d58b). Does this address your concerns?